monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/DoOnSubscribeObservable.scala

```diff
@@ -103,7 +103,7 @@ private[reactive] object DoOnSubscribeObservable {
       )
 
       val ref = SingleAssignCancelable()
-      val conn = StackedCancelable(List(cancelable, ref))
+      val conn = StackedCancelable(List(ref, cancelable))
```

Comment:
It is popped when `task` finishes running but `StackedCancelable` works in FIFO order so it was removing `cancelable`